app/crud/base.py

```diff
@@ -0,0 +1,117 @@
+from contextlib import contextmanager
+from snowflake.snowpark import Row
+from pydantic import BaseModel
+from typing import ClassVar, get_args, get_origin, Union, Dict
+import datetime
+
+
+## TODO
+# same again for probes
+# hook up to stored procs for (CRUD and validation)
+class BaseOpsCenterModel(BaseModel):
+    # The name of the table in snowflake (without schema) that the model maps to.
+    table_name: ClassVar[str] = None
+
+    @classmethod
+    def cols_dict(cls) -> Dict[str, str]:
+        cols = dict()
+        for field, info in cls.__fields__.items():
+            if isinstance(type, type(info.annotation)):
+                info_type = handle_type(info.annotation)
+            else:
+                info_type = handle_union(
+                    get_args(info.annotation), get_origin(info.annotation)
+                )
+            cols[field] = info_type
+        return cols
+
+    @classmethod
+    def create_table(cls, session, with_catalog_view=True):
+        cols = cls.cols_dict()
+        cols_str = ", ".join([f"{k} {v}" for k, v in cols.items()])
+        session.sql(
+            f"CREATE TABLE IF NOT EXISTS internal.{cls.table_name} ({cols_str})"
+        ).collect()
+        if with_catalog_view:
+            session.sql(
+                f"CREATE OR REPLACE VIEW catalog.{cls.table_name} AS SELECT * FROM internal.{cls.table_name}"
+            ).collect()
+
+    def write(self, session):
+        df = session.create_dataframe([Row(**dict(self))])
+        df.write.mode("append").save_as_table(f"INTERNAL.{self.table_name}")
+
+    def get_id(self) -> str:
+        """
+        Returns the column value for the current row which is unique among all other rows.
+        :return:
+        """
+        return None
+
+    def get_id_col(self) -> str:
+        """
+        Returns the name of the column which is unique among all other rows.
+        :return:
+        """
+        return None
+
+    def delete(self, session):
+        session.sql(
+            f"DELETE FROM INTERNAL.{self.table_name} WHERE {self.get_id_col()} = ?",
+            params=(self.get_id(),),
+        ).collect()
+
+    def update(self, session, obj) -> "BaseOpsCenterModel":
+        cols = dict(obj)
+        # Build up the SET clause and bind param values
+        set_elements = []
+        params = []
+        for k, v in cols.items():
+            set_elements.append(f"{k} = ?")
+            params.append(v)
```

Comment:
I'm not sure why this is happening, but I get an error running inside of Snowflake but not when running locally. Feels like some difference in version of snowpark session API.

```
Numeric value 'None' is not recognized
File "/usr/lib/python_udf/7529d9b466ef006af9b8b8149d6967acf5f82bbeb9479eea064c6239760d9a0a/lib/python3.8/site-packages/streamlit/runtime/scriptrunner/script_runner.py", line 561, in _run_script
    self._session_state.on_script_will_rerun(rerun_data.widget_states)
File "/usr/lib/python_udf/7529d9b466ef006af9b8b8149d6967acf5f82bbeb9479eea064c6239760d9a0a/lib/python3.8/site-packages/streamlit/runtime/state/safe_session_state.py", line 68, in on_script_will_rerun
    self._state.on_script_will_rerun(latest_widget_states)
File "/usr/lib/python_udf/7529d9b466ef006af9b8b8149d6967acf5f82bbeb9479eea064c6239760d9a0a/lib/python3.8/site-packages/streamlit/runtime/state/session_state.py", line 482, in on_script_will_rerun
    self._call_callbacks()
File "/usr/lib/python_udf/7529d9b466ef006af9b8b8149d6967acf5f82bbeb9479eea064c6239760d9a0a/lib/python3.8/site-packages/streamlit/runtime/state/session_state.py", line 495, in _call_callbacks
    self._new_widget_state.call_callback(wid)
File "/usr/lib/python_udf/7529d9b466ef006af9b8b8149d6967acf5f82bbeb9479eea064c6239760d9a0a/lib/python3.8/site-packages/streamlit/runtime/state/session_state.py", line 247, in call_callback
    callback(*args, **kwargs)
File "/home/udf/27672521534107002/labels.py", line 236, in on_update_click
    _ = old_label.update(sf, new_label)
File "/home/udf/27672521534107002/crud.zip/crud/labels.py", line 108, in update
    super().update(session, obj)
File "/home/udf/27672521534107002/crud.zip/crud/base.py", line 74, in update
    session.sql(
...
```
feels like snowpark session is trying (and failing) to coerce `None` to an `int`. Updating a label is broken as a result. Trying to figure out what is happening.